File: gaiag/baseline/alarm.py
# Gaiag --- Guile in Asd In Asd in Guile.
# Copyright © 2014 Jan Nieuwenhuizen <janneke@gnu.org>
#
# This file is part of Gaiag.
#
# Gaiag is free software: you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# Gaiag is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public
# License along with Gaiag.  If not, see <http://www.gnu.org/licenses/>.
# 
# Commentary:
# 
# Code:

import collections
import inspect
import sys

def __info__ (depth=1):
    return inspect.stack ()[depth]

def __file_name__ (depth=1):
    return __info__ (depth + 1)[1]

def __line__ (depth=1):
    return __info__ (depth + 1)[2]

def __function__ (depth=1):
    return __info__ (depth + 1)[3]

def object_bind (object, function):
    return lambda *args: function (object, *args)

class Console ():
    def __init__ (self):
        self._in = collections.namedtuple ('arm', 'disarm')
        self._out = collections.namedtuple ('detected', 'deactivated')

class Sensor ():
    def __init__ (self):
        self._in = collections.namedtuple ('enable', 'disable')
        self._out = collections.namedtuple ('triggered', 'disabled')

class Siren ():
    def __init__ (self):
        class In ():
            self.turnon = None
            self.turnoff = None
        self._in = In ()
        self._out = None

class Alarm ():
    def __init__ (self):
        self.state = 'disarmed'
        self.sounding = False
        self.po_console = Console ()
        self.po_sensor = Sensor ()
        self.po_siren = Siren ()

        #self.po_console._in.arm = object_bind (self, self.po_console_arm)
        self.po_console._in.arm = self.po_console_arm
        self.po_console._in.disarm = self.po_console_disarm
        self.po_sensor._out.triggered = self.po_sensor_triggered
        self.po_sensor._out.disabled = self.po_sensor_disabled

    def po_console_arm (self):
        sys.stderr.write (self.__class__.__name__ + '.' + __function__ () + '\n')

    def po_console_disarm (self):
        sys.stderr.write (self.__class__.__name__ + '.' + __function__ () + '\n')

    def po_sensor_triggered (self):
        sys.stderr.write (self.__class__.__name__ + '.' + __function__ () + '\n')

    def po_sensor_disabled (self):
        sys.stderr.write (self.__class__.__name__ + '.' + __function__ () + '\n')

class AlarmSystem ():
    def __init__ (self):
        self.is_alarm = Alarm ()
        self.is_sensor = Sensor ()
        self.is_siren = Siren ()
        self.po_console = self.is_alarm.po_console
        self.is_sensor.po_sensor = self.is_alarm.po_sensor
        self.is_siren.po_siren = self.is_alarm.po_siren

def detected ():
   sys.stderr.write ('Console.detected\n')

def deactivated ():
   sys.stderr.write ('Console.deactivated\n')

def main ():
    alarm_system = AlarmSystem ()
    alarm_system.po_console._out.detected = detected
    alarm_system.po_console._out.deactivated = deactivated

    alarm_system.po_console._in.arm ()
    alarm_system.is_sensor.po_sensor._out.triggered ()
    alarm_system.po_console._in.disarm ()
    alarm_system.is_sensor.po_sensor._out.disabled ()
    print alarm_system

if __name__ == '__main__':
    main ()

File: gaiag/baseline/Alarm.java
// Dezyne --- Dezyne command line tools
//
// Copyright © 2014, 2015 Jan Nieuwenhuizen <janneke@gnu.org>
// Copyright © 2014 Rutger van Beusekom <rutger.van.beusekom@verum.com>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

class Alarm extends Component {
  enum States {
    Disarmed, Armed, Triggered, Disarming
  };

  States state;
  Boolean sounding;

  IConsole console;
  ISensor sensor;
  ISiren siren;

  public Alarm(Runtime runtime) {this(runtime, "");};

  public Alarm(Runtime runtime, String name) {this(runtime, name, null);};

  public Alarm(Runtime runtime, String name, SystemComponent parent) {
    super(runtime, name, parent);
    this.flushes = true;
    state = States.Disarmed;
    sounding = false;
    console = new IConsole();
    console.in.name = "console";
    console.in.self = this;
    state = States.Disarmed;
    sounding = false;
    sensor = new ISensor();
    sensor.out.name = "sensor";
    sensor.out.self = this;
    siren = new ISiren();
    siren.out.name = "siren";
    siren.out.self = this;
    console.in.arm = new Action() {public void action() {Runtime.callIn(Alarm.this, new Action() {public void action() {console_arm();}}, new Meta(Alarm.this.console, "arm"));};};

    console.in.disarm = new Action() {public void action() {Runtime.callIn(Alarm.this, new Action() {public void action() {console_disarm();}}, new Meta(Alarm.this.console, "disarm"));};};

    sensor.out.triggered = new Action() {public void action() {Runtime.callOut(Alarm.this, new Action() {public void action() {sensor_triggered();}}, new Meta(Alarm.this.sensor, "triggered"));};};

    sensor.out.disabled = new Action() {public void action() {Runtime.callOut(Alarm.this, new Action() {public void action() {sensor_disabled();}}, new Meta(Alarm.this.sensor, "disabled"));};};

  };
  public void console_arm() {
    if (state == States.Disarmed) {
      {
        sensor.in.enable.action();
        state = States.Armed;
      }
    }
    else if (state == States.Armed) {
      throw new RuntimeException("illegal");
    }
    else if (state == States.Disarming) {
      throw new RuntimeException("illegal");
    }
    else if (state == States.Triggered) {
      throw new RuntimeException("illegal");
    }
  };

  public void console_disarm() {
    if (state == States.Disarmed) {
      throw new RuntimeException("illegal");
    }
    else if (state == States.Armed) {
      {
        sensor.in.disable.action();
        state = States.Disarming;
      }
    }
    else if (state == States.Disarming) {
      throw new RuntimeException("illegal");
    }
    else if (state == States.Triggered) {
      {
        sensor.in.disable.action();
        siren.in.turnoff.action();
        sounding = false;
        state = States.Disarming;
      }
    }
  };

  public void sensor_triggered() {
    if (state == States.Disarmed) {
      throw new RuntimeException("illegal");
    }
    else if (state == States.Armed) {
      {
        console.out.detected.action();
        siren.in.turnon.action();
        sounding = true;
        state = States.Triggered;
      }
    }
    else if (state == States.Disarming) {
      { }
    }
    else if (state == States.Triggered) {
      throw new RuntimeException("illegal");
    }
  };

  public void sensor_disabled() {
    if (state == States.Disarmed) {
      throw new RuntimeException("illegal");
    }
    else if (state == States.Armed) {
      throw new RuntimeException("illegal");
    }
    else if (state == States.Disarming && sounding) {
      console.out.deactivated.action();
      siren.in.turnoff.action();
      state = States.Disarmed;
      sounding = false;
    }
    else if (state == States.Disarming && ! (sounding)) {
      console.out.deactivated.action();
      state = States.Disarmed;
    }
    else if (state == States.Triggered) {
      throw new RuntimeException("illegal");
    }
  };

}
